api/Services/CarParkService.js

```diff
@@ -1,5 +1,66 @@
-let getAllCarParks = async (collection) => {
+const NearestHourService = require('./NearestHourService')
+
+const getAllCarParks = async (collection) => {
     return collection.find({}).toArray()
 }
 
-module.exports.getAllCarParks = getAllCarParks
\ No newline at end of file
+const getAvailableCarParks = async (carParkCollection, duration) => {
+    const secsInHour = 60 * 60
+    const startTime = NearestHourService()
+    const endTime = startTime + (duration * secsInHour)
+    let carParks = await carParkCollection.aggregate([
+        {
+            $lookup: {
+                from: 'bookings',
+                localField: '_id',
+                foreignField: 'carParkId',
+                as: 'bookings'
+            }
+        },
+        {
+            $addFields: {
+                bookings: {
+                    $filter: {
+                        input: '$bookings',
+                        as: 'bookings',
+                        cond: {
+                            $and: [
+                                {
+                                    $lt: ['$$bookings.startDateTime', endTime]
+                                },
+                                {
+                                    $gt: [{
+                                        $sum: ['$$bookings.startDateTime',
+                                            {$multiply: ['$$bookings.duration', secsInHour]}]
+                                    },
+                                        startTime]
+                                }
+                            ]
+                        }
+                    }
+                }
+            }
+        },
+        {
+            $addFields: {
+                availableSpaces: {
+                    $subtract: ['$totalSpaces', {
+                        $size: '$bookings'
+                    }]
+                }
+            }
+        },
+        {
+            $unset: 'bookings'
+        }
+
+
+    ]).toArray()
```

Comment:
Woah this is one hell of a query! It doesn't quite seem to be working as intended based on my testing with the Avon Street parking site.

I modified the existing booking for avon street so it's start time is 20:00 this evening (1623870000) and ran a search for current availability with various durations, and it seems that the existence of a booking for this site stops it from appearing in the results, even if the booking is well in the future or even in the past.

Reply:
Worked when deleted the second avon booking (ie the 4th test booking datum)